org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -147,6 +147,11 @@ public boolean finishConnect() throws IOException {
                 } else if (state == SSLEngineResult.HandshakeStatus.NEED_UNWRAP) {
                     // Read into an empty buffer to unwrap.
                     read(empty);
+                } else if (state == SSLEngineResult.HandshakeStatus.NEED_TASK) {
+                    // Run the task, synchronously, because we're a mostly
+                    // blocking SSLSocket.
+                    Runnable task = engine.getDelegatedTask();
+                    task.run();
```

Comment:
It is in a bigger loop. At each step, we either need `TASK`, `WRAP`, or `UNWRAP`. We run one task or perform one wrap at a time and step the outer loop. We don't run multiple tasks, and we leave it to `read` and `write` to handle the semantics of `unwrap` and `wrap` correctly. We could run it in a smaller loop, but why? The outer loop will suffice. :)

---

Also, `JSSSocketChannel` really only works with our particular `SSLEngine`. We have a single `Runnable`, which does all certificate validation (from all `X509TrustManagers`) at once. We don't have multiple `Runnable` tasks, in part because they can only happen during the handshake (and most of the handshake occurs inside NSS) -- not during the data transmission part.

Reply:
As an aside, the reason why we have the inner loop in `TestSSLEngine` is because we're trying to get all client work done _before_ bothering the server, and then again, get all _server_ work done before trying to bother the _client_, in the next loop iteration. There's one outer loop for trying to make the two talk. Here, the other client is on the other side of the socket, so we just need one loop, to make this client talk.

Reply:
Sounds good.